### PR TITLE
Add metadata hostcalls to `http_downstream`

### PR DESCRIPTION
**Pull request by ulyssa**
This PR adds the new downstream metadata hostcalls that have request handle arguments as part of the reusable sessions work.

I have also updated several of the downstream metadata hostcalls that currently return `Error::NotAvailable` to return `Error::ValueAbsent` instead so that they will actually return `None` in the SDK instead, which should help consumers with writing local tests that use those request metadata methods. (I have also updated the tests to demonstrate that they now return `None`.)

### CODE REVIEW

**Review by cceckman-at-fastly — DISMISSED** [review]
Some nits on docstrings but otherwise LGTM!

File: lib/src/session.rs [hunk 1]
```diff
@@ -1115,8 +1155,8 @@ impl Session {
     }
 
     /// Returns the unique identifier for the request this session is processing.
```
[the remark this answers is not in the document]
Comment (ulyssa) [reply lib/src/session.rs]: Done in b0e68e04c6df6bb8ec37623c01dcca7351dbb763
